### DISCUSSION TITLE
Support with Azure Cosmos DB storage

### DISCUSSION THREAD

**Discussion by deepanjan0604** [jobrunr/jobrunr#950] [Q&A]
Getting this error after migrating from MongoDB to Cosmos DB, I tried creating the indexed for jobs and recurring_jobs collection but still query is failing.

Error notifying JobStorageChangeListeners
com.mongodb.MongoQueryException: Query failed with error code 2 and error message 'Error=2, Details='Response status code does not indicate success: BadRequest (400); Substatus: 0; ActivityId: 8760e7f9-3e41-4d5c-8aa3-8472e4fb2e6b; Reason: (Response status code does not indicate success: BadRequest (400); Substatus: 0; ActivityId: 8760e7f9-3e41-4d5c-8aa3-8472e4fb2e6b; Reason: (Response status code does not indicate success: BadRequest (400); Substatus: 0; ActivityId: 8760e7f9-3e41-4d5c-8aa3-8472e4fb2e6b; Reason: (Message: {"Errors":["The index path corresponding to the specified order-by item is excluded."]}
ActivityId: 8760e7f9-3e41-4d5c-8aa3-8472e4fb2e6b, Request URI: /apps/fbe73e29-08e7-435a-b7cd-7f771a12047b/services/d1343df4-0b59-45df-9850-7419439fe86b/partitions/fa122114-ad22-4cd8-9c46-d17ad8993323/replicas/133424421053050582s/, RequestStats: Microsoft.Azure.Cosmos.Tracing.TraceData.ClientSideRequestStatisticsTraceDatum, SDK: Windows/10.0.20348 cosmos-netstandard-sdk/3.18.0);););' on server mongo.cosmos.azure.com:10255

**Comment by deepanjan0604** [comment]
Hi @rdehuyss can you please help here?

**Reply by rdehuyss** [reply]
I'm afraid we don't support cosmos db. 

For customers, we can take a look what the effort is to add support for it and do paid fracture development.

**Comment by deepanjan0604** [comment]
Hi @rdehuyss Can you provide some documentation on steps to migrate the storage provider from mongodb to SQL?

**Reply by rdehuyss** [reply]
I'm sorry but this is an opensource project and we don't have time to support the requests from every opensource user. 

Opensource != Free developers and support.

**Comment by limuelzoleta** [comment]
Hi @deepanjan0604  @rdehuyss  I recently worked on this in CosmosDB, and the issue was caused by a missing index on the background_job_servers collection. Where it is trying to sort firstHeartbeat. the simple solution is to create an index on the firstHeartbeat field.

@rdehuyss  is this something that can be added when running the migrations for mongodb collections?

**Reply by rdehuyss** [reply]
Hi @limuelzoleta - thank you for the research! Can you open a PR for this?
